File: algorithm/003-longest_substring_without_repeating_characters/python/solution.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        alphabet = {}
        longest = 0
        i = 0
        j = 0

        for c in s:
            if c in alphabet:
                longest = max(longest, j - i)
                while i < alphabet[c]:
                    alphabet[s[i]] = 0
                    i += 1
            j += 1
            alphabet[c] = j

        return max(longest, j - i)

File: algorithm/003-longest_substring_without_repeating_characters/go/solution.go
package solution

func max(n1 int, n2 int) int {
	if n1 > n2 {
		return n1
	}
	return n2
}

func lengthOfLongestSubstring(s string) int {
	var alphabet = [127]int{0}
	var longest = 0
	var i = 0
	var j = 0

	for j = 0; j < len(s); j++ {
		var c = s[j]
		if alphabet[c] > 0 {
			longest = max(longest, j-i)
			for ; i < alphabet[c]; i++ {
				alphabet[s[i]] = 0
			}
		}

		alphabet[c] = j + 1
	}

	return max(longest, j-i)
}
